### PR TITLE
Draft EEP: Testing Embulk plugins with JUnit 5

### PR DESCRIPTION
**Pull request by dmikurube**
This draft EEP summarizes a design of the new Embulk plugin testing framework with JUnit 5.

It still has some missing links, such as:
* How to avoid requiring other Embulk plugins to test the target Embulk plugin?
* How to provide the test input, and check the results?
* ...

However, implementing it would need some time, and the design may change through implementation.

If the design so far seems okay with the Core Team, I wanted to merge this EEP still as "Draft", and to revisit it later along with implementation.

### CODE REVIEW

**Comment by dmikurube** [comment]
If this draft looks okay for @embulk/core-team, I'll be merging this to `master`. Please take a look.

File: docs/eeps/draft-new-plugin-tests-with-junit5.md [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    implementation "org.embulk:embulk-util-config:..."
+    // ...
+
+    // Simple unit tests.
+    testImplementation "org.junit.jupiter:junit-jupiter-api:..."
+    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine:..."
+
+    // "Integrated" Embulk plugin tests.
+    embulkPluginTestImplementation "org.junit.jupiter:junit-jupiter-api:..."
+
```
Comment (hiroyuki-sato) [inline docs/eeps/draft-new-plugin-tests-with-junit5.md:115]: Just question, Does "Integrated" Embulk plugin test use `org.junit.jupyter`? not `org.embulk..`?

File: docs/eeps/draft-new-plugin-tests-with-junit5.md [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+We have to catch up with JUnit 5, definitely, as mentioned in the Motivation section.
+
+Other plugins
+--------------
+
+Testing with `embulk-junit4` often needed other plugins. For example, testing a File Input plugin (ex. S3 File Input) needed a CSV Parser plugin to confirm the input files are retrieved by the File Input plugin, and then processed in later stages properly.
+
+However, they are unnecessary dependencies, which could bring unexpected impacts when the depended plugin changes. The CSV Parser has changed its internal structure around v0.11.0, and it actually made some impacts.
+
```
Comment (hiroyuki-sato) [inline docs/eeps/draft-new-plugin-tests-with-junit5.md:45]: Does this mean If a plugin developer develop a FileInputPlugin, this test framework provides parser, decoder, and output plugin(like mock) for test?

**Review by hiroyuki-sato — APPROVED** [review]
LGTM👍 (I have two questions.)

File: docs/eeps/draft-new-plugin-tests-with-junit5.md (hunk 2 above)
Comment (dmikurube) [reply docs/eeps/draft-new-plugin-tests-with-junit5.md:45]: That's the point of "How to provide the test input, and check the results?" :)

I now think "other plugins" are not needed, but only inputs and methods to check the results would enable testing.

For example for File Input plugins, just testing the "read" Buffer should be sufficient, but no need to paese it.

File: docs/eeps/draft-new-plugin-tests-with-junit5.md (hunk 1 above)
Comment (dmikurube) [reply docs/eeps/draft-new-plugin-tests-with-junit5.md:115]: That's another point I still wonder. However, without jupiter-api, we'll have to implement so many things, including all `assert`s and else. I expect we'll want at least jupiter-api...

**Review by dmikurube — COMMENTED** [review]
Thanks for your comments! Replied in-line.

**Comment by hiroyuki-sato** [comment]
@dmikurube Thanks! My question has been cleared up. 👍

**Comment by dmikurube** [comment]
Thanks! I'd like to update this EEP later with the information in early new year.

**Pull request merged**